packages/pyright-internal/src/parser/tokenCollection.ts

```diff
@@ -0,0 +1,192 @@
+/*
+ * tokenCollection.ts
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ *
+ * Stores a stream of tokens from a python file.
+ */
+
+import { assert } from '../common/debug';
+import { TextRangeCollection } from '../common/textRangeCollection';
+import { Token } from './tokenizerTypes';
+
+// Interface implemented by a TokenCollection
+interface ITokenCollection {
+    start: number;
+    end: number;
+    length: number;
+    count: number;
+    contains(position: number): boolean;
+    getItemAt(index: number): Token;
+    getItemAtPosition(position: number): number;
+}
+
+class TokenCollectionFast extends TextRangeCollection<Token> implements ITokenCollection {
+    constructor(private readonly _tokens: Token[]) {
+        super(_tokens);
+    }
+
+    toArray(): Token[] {
+        return this._tokens;
+    }
+}
+
+function* generateCompressedData(
```

Comment:
I used a generator here because in order to generate an Int32Array, you need to either give it the initial size (which would require two passes through the token list) or give it an iterator.